doc/devdocs/offset-arrays.rst

```diff
@@ -0,0 +1,301 @@
+array-types are expected to be supplied through packages.
+
+The purpose of this page is to address the question, "what do I
+have to do to support such arrays in my own code?"  First, let's
+address the simplest case: if you know that your code will never need
+to handle arrays with unconventional indexing, hopefully the answer is
+"nothing." Old code, on conventional arrays, should function
+essentially without alteration as long as it was using the exported
+interfaces of Julia.
+
+Generalizing existing code
+--------------------------
+
+As an overview, the steps are:
+
+- replace many uses of ``size`` with ``indices``
+- replace ``1:length(A)`` with ``linearindices(A)``
+- replace explicit allocations like ``Array{Int}(size(B))`` with ``similar(Array{Int}, indices(B))``
+- if necessary, wrap functions with ``@safeindices``
+
+These are described in more detail below.
+
+Safety features: new errors and ``@safeindices``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Because unconventional indexing breaks deeply-held assumptions
+throughout the Julia ecosystem, early-adopters running code that has
+not been updated are likely to experience errors.  The most
+frustrating bugs would be incorrect results or segfaults (total
+crashes of Julia).  For example, consider the following function::
+
+  function mycopy!(dest::AbstractVector, src::AbstractVector)
+      length(dest) == length(src) || throw(DimensionMismatch("vectors must match"))
+      # OK, now we're safe to use @inbounds, right? (not anymore!)
+      for i = 1:length(src)
+          @inbounds dest[i] = src[i]
+      end
+      dest
+  end
+
+This code implicitly assumes that vectors are indexed from 1.
+Previously that was a safe assumption, so this code was fine, but
+(depending on what types the user passes to this function) it may no
+longer be safe.  If this code continued to work when passed a vector
+with non-1 indices, it would either produce an incorrect answer or it
+would segfault.  (If you do get segfaults, to help locate the
+cause try running julia with the option ``--check-bounds=yes``.)
+
+To ensure that such errors are caught, in julia-0.5 both ``length``
+and ``size`` throw an error (by default) when passed an array with
+non-1 indexing.  This is designed to force users of such arrays to
+check the code, and inspect it for whether it needs to be generalized.
+Once the code has been modified (see below), if needed you can wrap
+this function using the macro ``@safeindices``.  For example::
+
+  @safeindices longer_than_5(A::AbstractArray) = length(A) > 5
+
+would ensure that this call to ``length`` succeeds even if ``A`` has
+non-1 indices.
+
+In many cases, you may find that you eliminate ``size`` and ``length``
+from functions by replacing them with ``indices`` and
+``linearindices``, so ``@safeindices`` is not needed in every case.
+
+You can read about how ``@safeindices`` works in the on-line help.
```

Comment:
"online help" or a cross-reference to the stdlib docs if that works from devdocs